### PR TITLE
fix(preview-generator): simplified and typed code

### PR DESCRIPTION
**Pull request by Galitan-dev**
# fix(preview-generator): simplified and typed code

After seeing @bdebon's video about the project, I looked at the preview generator and as he said, it's a little messy. So I made what I can.

## Files

The affected files are:
- preview-generator.ts 
- generator-preview-utils/*
- preview images: the title is a bit more centered

## Fixes

- Typed some variables and arguments
- Removed duplicate code
- Usage of javascript's default argument feature
- Absolute path
- Some code simplification

You can read more in my code review

### CODE REVIEW

File: generator-preview-utils/wrap-text-canvas.ts [hunk 1]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
   lineHeight: number
 ) {
   // First, start by splitting all of our text into words, but splitting it into an array split by spaces
-  let words = text.split(' ')
+  let words = text.split(/\s+/)
   let line = '' // This will store the text of the current line
-  let testLine = '' // This will store the text when we add a word, to test if it's too long
-  let lineArray = [] // This is an array of lines, which the function will return
+  let lineArray: [string, number, number][] = [] // This is an array of lines, which the function will return
+                                                 // [text, x, y]
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:21]: typed `lineArray` (+ explanations)

File: generator-preview-utils/wrap-text-canvas.ts [hunk 2]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
   lineHeight: number
 ) {
   // First, start by splitting all of our text into words, but splitting it into an array split by spaces
-  let words = text.split(' ')
+  let words = text.split(/\s+/)
   let line = '' // This will store the text of the current line
-  let testLine = '' // This will store the text when we add a word, to test if it's too long
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:20]: It's simpler with one variable, as they do quiet the same

File: generator-preview-utils/wrap-text-canvas.ts [hunk 3]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
...
-  let lineArray = [] // This is an array of lines, which the function will return
+  let lineArray: [string, number, number][] = [] // This is an array of lines, which the function will return
+                                                 // [text, x, y] 
 
   // Lets iterate over each word
-  for (var n = 0; n < words.length; n++) {
+  for (const word of words) {
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:24]: `n` was not required anymore

File: generator-preview-utils/wrap-text-canvas.ts [hunk 4]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
...
     // Create a test line, and measure it..
-    testLine += `${words[n]} `
-    let metrics = ctx.measureText(testLine)
+    let metrics = ctx.measureText(line + ' ' + word)
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:26]: the only usage of testLine

File: generator-preview-utils/wrap-text-canvas.ts [hunk 5]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
...
     let testWidth = metrics.width
-    // If the width of this test line is more than the max width
-    if (testWidth > maxWidth && n > 0) {
+    // If the width of this line is more than the max width
+    if (testWidth > maxWidth && ctx.measureText(word).width <= maxWidth) {
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:29]: the issue of words bigger than the maxWidth does not only occurs for the first words

File: generator-preview-utils/wrap-text-canvas.ts [hunk 6]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
...
-      line = `${words[n]} `
-      testLine = `${words[n]} `
+      // Update line to use this word as the first word on the next line
+      line = word
     } else {
       // If the test line is still less than the max width, then add the word to the current line
-      line += `${words[n]} `
-    }
-    // If we never reach the full max width, then there is only one line.. so push it into the lineArray so we return something
-    if (n === words.length - 1) {
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:42]: no need to check it for each word

File: generator-preview-utils/wrap-text-canvas.ts [hunk 7]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
...
+      line += ' ' + word
     }
   }
 
-  const heightText = lineArray.length * lineHeight
-  let firstY = maxHeight / 2 - heightText / 2 + lineHeight
+  // Push the last words
+  lineArray.push([line, x, 0])
+
+  const textHeight = lineArray.length * lineHeight
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:43]: `heightText` -> `textHeight`,
like `lineHeight`

File: generator-preview-utils/wrap-text-canvas.ts [hunk 8]
```diff
@@ -15,67 +15,41 @@ export const wrapText = function (
...
-      line += `${words[n]} `
-    }
-    // If we never reach the full max width, then there is only one line.. so push it into the lineArray so we return something
-    if (n === words.length - 1) {
-      lineArray.push([line, x, y])
-    }
+  for (const line of lineArray) {
+    line[0]
+    line[2] = y;
+    y += lineHeight;
```
Comment (Galitan-dev) [inline generator-preview-utils/wrap-text-canvas.ts:50]: we only need to calculate `y` based on the numbers of line.
Why recalculate all lines ?

File: preview-generator.ts [hunk 9]
```diff
@@ -1,17 +1,18 @@
...
 import { createTextWithBackground, drawImageProp } from './generator-preview-utils/utils'
-import * as fs from 'fs'
 import { questions } from './apps/devchoices-next/public/assets/data/questions'
+import { Choice, QuestionInterface } from './libs/shared/interfaces/src'
+import { createWriteStream, existsSync } from 'fs'
 
 const width = 1200
 const height = 628
-const assetPath = 'apps/devchoices-next/public'
+const assetPath = `${__dirname}/apps/devchoices-next/public`
```
Comment (Galitan-dev) [inline preview-generator.ts:9]: absolute path

File: preview-generator.ts [hunk 10]
```diff
@@ -1,17 +1,18 @@
 import { Canvas, Image, loadImage } from 'canvas'
 import { createTextWithBackground, drawImageProp } from './generator-preview-utils/utils'
-import * as fs from 'fs'
 import { questions } from './apps/devchoices-next/public/assets/data/questions'
+import { Choice, QuestionInterface } from './libs/shared/interfaces/src'
+import { createWriteStream, existsSync } from 'fs'
```
Comment (Galitan-dev) [inline preview-generator.ts:5]: avoiding `*` imports

File: preview-generator.ts [hunk 11]
```diff
@@ -1,17 +1,18 @@
...
 const height = 628
-const assetPath = 'apps/devchoices-next/public'
+const assetPath = `${__dirname}/apps/devchoices-next/public`
 
 // Checks for --override and if it has a value
 const override = Boolean(process.argv.indexOf('--override') > -1)
 
-const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question: any) => {
-  if (!question) return
+const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, choice: Choice) => {
```
Comment (Galitan-dev) [inline preview-generator.ts:14]: we don't need the whole question, but only the choice

File: preview-generator.ts [hunk 12]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
     ctx,
     width,
     height
   )
 }
 
-const renderPreviewBySlug = async (slug: string, override = false) => {
-  const question = questions.find((q) => q.slug === slug)
-  if (!question) return
+const renderPreview = async (question: QuestionInterface, override = false) => {
```
Comment (Galitan-dev) [inline preview-generator.ts:33]: Oh, another arrow function. I didn't see it

File: preview-generator.ts [hunk 13]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
 
-  if (fs.existsSync(`${assetPath}/assets/img-previews/preview-${question.slug}.jpg`) && !override) {
+  if (!override && existsSync(`${assetPath}/assets/img-previews/preview-${question.slug}.jpg`)) {
```
Comment (Galitan-dev) [inline preview-generator.ts:35]: optimization

File: preview-generator.ts [hunk 14]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
     console.log(`Preview for ${question.slug} already exists`)
     return
   }
 
-  if (question.choiceLeft.img_path.indexOf('webp') > -1 || question.choiceRight.img_path.indexOf('webp') > -1) {
+  if (question.choiceLeft.img_path.endsWith('.webp') || question.choiceRight.img_path.endsWith('webp')) {
```
Comment (Galitan-dev) [inline preview-generator.ts:40]: Imagine we add a question like "webp or png ?"

File: preview-generator.ts [hunk 15]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
     console.log('Webp not supported')
     return
   }
 
-  const promiseImg1 = loadImage(assetPath + question.choiceLeft.img_path)
-  const promiseImg2 = loadImage(assetPath + question.choiceRight.img_path)
-  Promise.all([promiseImg1, promiseImg2]).then(async (images) => {
+  const leftImagePromise = loadImage(assetPath + question.choiceLeft.img_path)
```
Comment (Galitan-dev) [inline preview-generator.ts:45]: a more precise variable name

File: preview-generator.ts [hunk 16]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
     const canvas = new Canvas(width, height)
     const ctx = canvas.getContext('2d')
 
-    await drawSide('left', ctx, images[0], question)
-    await drawSide('right', ctx, images[1], question)
+    await drawSide('left', ctx, leftImage, question.choiceLeft)
+    await drawSide('right', ctx, rightImage, question.choiceRight)
 
-    const out = fs.createWriteStream(__dirname + `/${assetPath}/assets/img-previews/preview-${slug}.jpg`)
+    const out = createWriteStream(`${assetPath}/assets/img-previews/preview-${question.slug}.jpg`)
```
Comment (Galitan-dev) [inline preview-generator.ts:54]: Use `` everywhere for string concatenation

File: preview-generator.ts [hunk 17]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
+    out.on('finish', () => console.log(`The JPG preview-${question.slug}.jpg file was created.`))
   })
 }
 
-//renderPreviewBySlug('camelCase-or-snake_case', true).then(() => {})
-
-questions.forEach(async (question) => {
-  await renderPreviewBySlug(question.slug, override)
+questions.forEach((question) => {
+  renderPreview(question, override)
```
Comment (Galitan-dev) [inline preview-generator.ts:64]: we don't have things to do after `renderPreview` and node will always wait for the entire code to run. So no need of `await`

File: preview-generator.ts [hunk 18]
```diff
@@ -21,48 +22,44 @@ const drawSide = async (side: 'left' | 'right', ctx: any, image: Image, question
...
     const stream = canvas.createJPEGStream({
       quality: 0.8,
     })
     stream.pipe(out)
-    out.on('finish', () => console.log('The JPG preview-' + slug + '.jpg file was created.'))
+    out.on('finish', () => console.log(`The JPG preview-${question.slug}.jpg file was created.`))
   })
 }
 
-//renderPreviewBySlug('camelCase-or-snake_case', true).then(() => {})
```
Comment (Galitan-dev) [inline preview-generator.ts:64]: intense debugging!

**Review by Galitan-dev — COMMENTED** [review]
This changes are just based on my preferences, so it would be normal that you don't agree with some. Don't hesitate to request changes.

**Comment by Galitan-dev** [comment]
I cannot add labels to the pull request

**Comment by Drarig29** [comment]
We ended up doing a lot of similar things haha: https://github.com/bdebon/choiceof.dev/pull/114

**Comment by Galitan-dev** [comment]
> We ended up doing a lot of similar things haha: #114

Oh yes, but you could have watched other pr before :)

**Pull request closed without merging**